From eb00be9a68fd1ba9b224f6678b2b934dedb592d6 Mon Sep 17 00:00:00 2001
From: Yuvarani V <quic_yuvarani@quicinc.com>
Date: Fri, 24 Nov 2023 19:21:22 +0530
Subject: [PATCH] mac80211/cfg80211: Parse NL attribute to set critical update
 params

Parse NL attribute NL80211_ATTR_SET_CRITICAL_UPDATE to get critical
update params and pass this data to driver to update on beacon
template

Signed-off-by: Rathees Kumar R Chinannan <quic_rrchinan@quicinc.com>
Signed-off-by: Yuvarani V <quic_yuvarani@quicinc.com>
---
 include/linux/ieee80211.h    | 66 ------------------------------------
 include/net/cfg80211.h       |  6 ++++
 include/net/mac80211.h       |  8 ++---
 include/uapi/linux/nl80211.h | 20 +++++++++++
 net/mac80211/cfg.c           | 14 ++++++++
 net/wireless/nl80211.c       | 34 +++++++++++++++++++
 6 files changed, 78 insertions(+), 70 deletions(-)

--- a/include/linux/ieee80211.h
+++ b/include/linux/ieee80211.h
@@ -5167,70 +5167,4 @@ static inline bool ieee80211_mle_sta_pro
 				 _data + ieee80211_mle_common_size(_data),\
 				 _len - ieee80211_mle_common_size(_data))
 
-/**
- * enum ieee80211_critical_updates - Critical Update (CU) flags
- *
- * These flags are used to indicate the type of critical update happening
- * on a link in an interface.
- *
- * @IEEE80211_CU_INCLUDE_CSA_ELEM: critical update due to inclusion of a Channel
- *	Switch Announcement element.
- * @IEEE80211_CU_INCLUDE_ECSA_ELEM: critical update due to inclusion of an
- *	Extended Channel Switch Announcement element.
- * @IEEE80211_CU_MODIFY_EDCA_PARAM_ELEM: critical update due to modification of
- *	the EDCA parameters element.
- * @IEEE80211_CU_INCLUDE_QUIET_ELEM: critical update due to inclusion of a Quiet
- *	element.
- * @IEEE80211_CU_MODIFY_DSSS_PARAM_ELEM: critical update due to modification of
- *	the DSSS Parameter Set.
- * @IEEE80211_CU_MODIFY_HT_OPER_ELEM: critical update due to modification of the
- *	HT Operation element
- * @IEEE80211_CU_INCLUDE_WBCS_ELEM: critical update due to inclusion of a Wide
- *	Bandwidth Channel Switch element.
- * @IEEE80211_CU_INCLUDE_CSW_ELEM: critical update due to inclusion of a Channel
- *	Switch Wrapper element.
- * @IEEE80211_CU_INCLUDE_OMN_ELEM: critical update due to inclusion of an
- *	Operating Mode Notification element.
- * @IEEE80211_CU_INCLUDE_Q_CHAN_ELEM: critical update due to inclusion of a
- *	Quiet Channel element.
- * @IEEE80211_CU_MODIFY_VHT_OPER_ELEM: critical update due to modification of the
- *	VHT Operation element.
- * @IEEE80211_CU_MODIFY_HE_OPER_ELEM: critical update due to modification of the
- *	HE Operation element.
- * @IEEE80211_CU_INCLUDE_B_TWT_ELEM: critical update due to inclusion a
- *	Broadcast TWT element.
- * @IEEE80211_CU_INCLUDE_B_TWT_EXIST_ELEM: critical update due to inclusion of a
- *	Broadcast TWT Parameter Set field in an existing Broadcast TWT element.
- * @IEEE80211_CU_INCLUDE_BCCA_ELEM: critical update due to inclusion of the BSS
- *	Color Change Announcement element.
- * @IEEE80211_CU_MODIFY_MU_EDCA_PARAM_ELEM: critical update due to modification
- *	of the MU EDCA Parameter Set element.
- * @IEEE80211_CU_MODIFY_SR_PARAM_ELEM: critical update due to modification of the
- *	Spatial Reuse Parameter Set element.
- * @IEEE80211_CU_MODIFY_UORA_PARAM_ELEM: critical update due to modification of
- *	the UORA Parameter Set element.
- * @IEEE80211_CU_MODIFY_EHT_OPER_ELEM: critical update due to modification of the
- *	EHT Operation element.
- */
-enum ieee80211_critical_updates {
-	IEEE80211_CU_INCLUDE_CSA_ELEM			= 1 << 0,
-	IEEE80211_CU_INCLUDE_ECSA_ELEM			= 1 << 1,
-	IEEE80211_CU_MODIFY_EDCA_PARAM_ELEM		= 1 << 2,
-	IEEE80211_CU_INCLUDE_QUIET_ELEM			= 1 << 3,
-	IEEE80211_CU_MODIFY_DSSS_PARAM_ELEM		= 1 << 4,
-	IEEE80211_CU_MODIFY_HT_OPER_ELEM		= 1 << 5,
-	IEEE80211_CU_INCLUDE_WBCS_ELEM			= 1 << 6,
-	IEEE80211_CU_INCLUDE_CSW_ELEM			= 1 << 7,
-	IEEE80211_CU_INCLUDE_OMN_ELEM			= 1 << 8,
-	IEEE80211_CU_INCLUDE_Q_CHAN_ELEM		= 1 << 9,
-	IEEE80211_CU_MODIFY_VHT_OPER_ELEM		= 1 << 10,
-	IEEE80211_CU_MODIFY_HE_OPER_ELEM		= 1 << 11,
-	IEEE80211_CU_INCLUDE_B_TWT_ELEM			= 1 << 12,
-	IEEE80211_CU_INCLUDE_B_TWT_EXIST_ELEM		= 1 << 13,
-	IEEE80211_CU_INCLUDE_BCCA_ELEM			= 1 << 14,
-	IEEE80211_CU_MODIFY_MU_EDCA_PARAM_ELEM		= 1 << 15,
-	IEEE80211_CU_MODIFY_SR_PARAM_ELEM		= 1 << 16,
-	IEEE80211_CU_MODIFY_UORA_PARAM_ELEM		= 1 << 17,
-	IEEE80211_CU_MODIFY_EHT_OPER_ELEM		= 1 << 18,
-};
 #endif /* LINUX_IEEE80211_H */
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -1485,6 +1485,8 @@ struct cfg80211_unsol_bcast_probe_resp {
  *	a 20 MHz channel, lowest bit corresponding to the lowest channel.
  *	Bit set to 1 indicates that the channel is punctured.
  * @beacon_tx_mode: Beacon Tx Mode setting
+ * @elemid_added: critical update element id inserted
+ * @elemid_modified: critical update element id modified
  */
 struct cfg80211_ap_settings {
 	struct cfg80211_chan_def chandef;
@@ -1523,6 +1525,8 @@ struct cfg80211_ap_settings {
 	enum nl80211_beacon_tx_mode beacon_tx_mode;
 	bool ap_ps_valid;
 	bool ap_ps_enable;
+	bool elemid_added;
+	bool elemid_modified;
 };
 
 /**
@@ -1558,6 +1562,8 @@ struct cfg80211_csa_settings {
 	u8 count;
 	u16 punct_bitmap;
 	struct cfg80211_unsol_bcast_probe_resp unsol_bcast_probe_resp;
+	bool beacon_after_cu;
+	bool beacon_csa_cu;
 };
 
 /**
@@ -1581,6 +1587,7 @@ struct cfg80211_color_change_settings {
 	u8 count;
 	u8 color;
 	struct cfg80211_unsol_bcast_probe_resp unsol_bcast_probe_resp;
+	bool elemid_added;
 };
 
 /**
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -705,8 +705,8 @@ struct ieee80211_fils_discovery {
  *	beamformee
  * @nss_ap_isolate: Used for notifying the NSS host about AP isolate feature
  * @beacon_tx_mode: Beacon Tx Mode setting.
- * @critical_update_flag: indicates any Critical update going on in the BSS.
- *	see &enum ieee80211_critical_updates
+ * @elemid_added: critical update element id inserted
+ * @elemid_modified: critical update element id modified
  */
 struct ieee80211_bss_conf {
 	struct ieee80211_vif *vif;
@@ -814,8 +814,8 @@ struct ieee80211_bss_conf {
 	bool nss_offld_mesh_forward_enabled;
 	u32 nss_offld_mpath_refresh_time;
 
-	/* Critical Update flag*/
-	u32 critical_update_flag;
+	bool elemid_added;
+	bool elemid_modified;
 	u32 rts_threshold;
 };
 
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -41,6 +41,7 @@
  */
 
 #include <linux/types.h>
+#include <linux/kernel.h>
 
 #define NL80211_GENL_NAME "nl80211"
 
@@ -2880,6 +2881,9 @@ enum nl80211_commands {
  *	update for each MLD. In each nested item, it contains attributes
  *	defined in &enum nl80211_cu_attrs.
  *
+ * @NL80211_ATTR_SET_CRITICAL_UPDATE: set critical update for the bss (u8)
+ *	(see &enum nl80211_set_cu).
+ *
  * @NUM_NL80211_ATTR: total number of nl80211_attrs available
  * @NL80211_ATTR_MAX: highest attribute number currently defined
  * @__NL80211_ATTR_AFTER_LAST: internal use
@@ -3447,6 +3451,7 @@ enum nl80211_attrs {
 	NL80211_ATTR_ADD_MULTI_CHAN,
 	NL80211_ATTR_DEL_MULTI_CHAN,
 	NL80211_ATTR_RXMGMT_CRITICAL_UPDATE,
+	NL80211_ATTR_SET_CRITICAL_UPDATE,
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
@@ -8211,4 +8216,20 @@ enum nl80211_cu_mld_link_attrs {
 	__NL80211_CU_MLD_LINK_ATTR_LAST,
 	NL80211_CU_MLD_LINK_ATTR_MAX = __NL80211_CU_MLD_LINK_ATTR_LAST - 1
 };
+
+/**
+ * enum nl80211_set_cu - set critical update params
+ *
+ * @NL80211_CU_UNSPEC: reserved
+ * @NL80211_CU_ELEMID_ADDED: Critical update - Element id added
+ * @NL80211_CU_ELEMID_MODIFIED: Critical update - Element id Modified
+ * @NUM_NL80211_CUS: number of CU attributes
+ */
+enum nl80211_set_cu {
+	NL80211_CU_UNSPEC = 0,
+	NL80211_CU_ELEMID_ADDED = BIT(0),
+	NL80211_CU_ELEMID_MODIFIED = BIT(1),
+
+	NUM_NL80211_CUS = BIT(2),
+};
 #endif /* __LINUX_NL80211_H */
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1651,6 +1651,8 @@ static int ieee80211_start_ap(struct wip
 		goto error;
 	changed |= BSS_CHANGED_UNSOL_BCAST_PROBE_RESP;
 
+	link_conf->elemid_added = params->elemid_added;
+	link_conf->elemid_modified = params->elemid_modified;
 	err = drv_start_ap(sdata->local, sdata, link_conf);
 	if (err) {
 		old = sdata_dereference(link->u.ap.beacon, sdata);
@@ -1659,6 +1661,8 @@ static int ieee80211_start_ap(struct wip
 			kfree_rcu(old, rcu_head);
 		RCU_INIT_POINTER(link->u.ap.beacon, NULL);
 		sdata->u.ap.active = false;
+		link_conf->elemid_added = 0;
+		link_conf->elemid_modified = 0;
 		goto error;
 	}
 
@@ -1670,6 +1674,8 @@ static int ieee80211_start_ap(struct wip
 	list_for_each_entry(vlan, &sdata->u.ap.vlans, u.vlan.list)
 		netif_carrier_on(vlan->dev);
 
+	link_conf->elemid_added = 0;
+	link_conf->elemid_modified = 0;
 	return 0;
 
 error:
@@ -1748,8 +1754,12 @@ static int ieee80211_update_ap(struct wi
 		return err;
 	changed |= BSS_CHANGED_UNSOL_BCAST_PROBE_RESP;
 
+	link_conf->elemid_added = params->elemid_added;
+	link_conf->elemid_modified = params->elemid_modified;
 
 	ieee80211_link_info_change_notify(sdata, link, changed);
+	link_conf->elemid_added = 0;
+	link_conf->elemid_modified = 0;
 	return 0;
 }
 
@@ -4129,6 +4139,7 @@ static int ieee80211_set_after_csa_beaco
 					  u64 *changed)
 {
 	struct ieee80211_sub_if_data *sdata = link->sdata;
+	struct ieee80211_bss_conf *link_conf = link->conf;
 	int err;
 
 	switch (sdata->vif.type) {
@@ -4144,9 +4155,8 @@ static int ieee80211_set_after_csa_beaco
 		if (err < 0)
 			return err;
 
-                link->conf->critical_update_flag &=
-					     ~(IEEE80211_CU_INCLUDE_CSA_ELEM |
-					       IEEE80211_CU_INCLUDE_ECSA_ELEM);
+		link_conf->elemid_modified = link->u.ap.after_beacon_cu;
+		link->u.ap.after_beacon_cu = 0;
 
 		break;
 	case NL80211_IFTYPE_ADHOC:
@@ -4275,6 +4285,7 @@ void ieee80211_csa_finalize_work(struct
 		goto unlock;
 
 	ieee80211_csa_finalize(link);
+	link_conf->elemid_modified = 0;
 
 unlock:
 	mutex_unlock(&local->chanctx_mtx);
@@ -4338,10 +4349,6 @@ static int ieee80211_set_csa_beacon(stru
 			return err;
 		}
 
-		link->conf->critical_update_flag |=
-					(IEEE80211_CU_INCLUDE_CSA_ELEM |
-					 IEEE80211_CU_INCLUDE_ECSA_ELEM);
-
 		break;
 	case NL80211_IFTYPE_ADHOC:
 		if (!sdata->vif.cfg.ibss_joined)
@@ -4492,6 +4499,8 @@ __ieee80211_channel_switch(struct wiphy
 	if (err)
 		goto out;
 	changed |= BSS_CHANGED_UNSOL_BCAST_PROBE_RESP;
+	link_conf->elemid_added = params->beacon_csa_cu;
+	link->u.ap.after_beacon_cu = params->beacon_after_cu;
 
 	/* should we pass link_id here ? */
 	err = drv_pre_channel_switch(sdata, &ch_switch);
@@ -4550,6 +4559,7 @@ __ieee80211_channel_switch(struct wiphy
 
 out:
 	mutex_unlock(&local->chanctx_mtx);
+	link_conf->elemid_added = 0;
 	return err;
 }
 
@@ -5290,8 +5300,6 @@ ieee80211_set_after_color_change_beacon(
 		if (ret < 0)
 			return ret;
 
-		link->conf->critical_update_flag &= ~IEEE80211_CU_INCLUDE_BCCA_ELEM;
-
 		break;
 	}
 	default:
@@ -5335,8 +5343,6 @@ ieee80211_set_color_change_beacon(struct
 			return err;
 		}
 
-		link->conf->critical_update_flag |= IEEE80211_CU_INCLUDE_BCCA_ELEM;
-
 		break;
 	default:
 		return -EOPNOTSUPP;
@@ -5578,6 +5584,7 @@ ieee80211_color_change(struct wiphy *wip
 
 	link_conf->color_change_active = true;
 	link_conf->color_change_color = params->color;
+	link_conf->elemid_added = params->elemid_added;
 
 	err = ieee80211_set_unsol_bcast_probe_resp(sdata, &params->unsol_bcast_probe_resp,
 						   link, link_conf);
@@ -5596,6 +5603,7 @@ ieee80211_color_change(struct wiphy *wip
 
 out:
 	mutex_unlock(&local->mtx);
+	link_conf->elemid_added = 0;
 
 	return err;
 }
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -887,6 +887,7 @@ static const struct nla_policy nl80211_p
 	[NL80211_ATTR_ADD_MULTI_CHAN] = { .type = NLA_U32 },
 	[NL80211_ATTR_DEL_MULTI_CHAN] = { .type = NLA_U32 },
 	[NL80211_ATTR_RXMGMT_CRITICAL_UPDATE] = NLA_POLICY_NESTED(cu_policy),
+	[NL80211_ATTR_SET_CRITICAL_UPDATE] = { .type = NLA_U8 },
 };
 
 /* policy for the key attributes */
@@ -6565,6 +6566,7 @@ static int nl80211_start_ap(struct sk_bu
 	struct wireless_dev *wdev = dev->ieee80211_ptr;
 	struct cfg80211_ap_settings *params;
 	int err;
+	u8 critical_update = 0;
 
 	if (dev->ieee80211_ptr->iftype != NL80211_IFTYPE_AP &&
 	    dev->ieee80211_ptr->iftype != NL80211_IFTYPE_P2P_GO)
@@ -6838,6 +6840,14 @@ static int nl80211_start_ap(struct sk_bu
 			info->attrs[NL80211_ATTR_AP_SETTINGS_FLAGS]);
 	else if (info->attrs[NL80211_ATTR_EXTERNAL_AUTH_SUPPORT])
 		params->flags |= NL80211_AP_SETTINGS_EXTERNAL_AUTH_SUPPORT;
+	if (info->attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]) {
+		critical_update =
+			nla_get_u8(info->attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]);
+		if (critical_update & NL80211_CU_ELEMID_ADDED)
+			params->elemid_added = 1;
+		if (critical_update & NL80211_CU_ELEMID_MODIFIED)
+			params->elemid_modified = 1;
+	}
 
 	if (wdev->conn_owner_nlportid &&
 	    info->attrs[NL80211_ATTR_SOCKET_OWNER] &&
@@ -6886,6 +6896,7 @@ static int nl80211_update_ap(struct sk_b
 	struct cfg80211_ap_settings *params;
 	int err;
 	bool haveinfo = false;
+	u8 critical_update = 0;
 
 	if (dev->ieee80211_ptr->iftype != NL80211_IFTYPE_AP &&
 	    dev->ieee80211_ptr->iftype != NL80211_IFTYPE_P2P_GO)
@@ -6928,6 +6939,15 @@ static int nl80211_update_ap(struct sk_b
 		if (err)
 			goto out;
 	}
+	if (info->attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]) {
+		critical_update =
+			nla_get_u8(info->attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]);
+		if (critical_update & NL80211_CU_ELEMID_ADDED)
+			params->elemid_added = 1;
+		if (critical_update & NL80211_CU_ELEMID_MODIFIED)
+			params->elemid_modified = 1;
+	}
+
 	wdev_lock(wdev);
 	err = rdev_update_ap(rdev, dev, params);
 	wdev_unlock(wdev);
@@ -10962,6 +10982,7 @@ static int nl80211_channel_switch(struct
 	bool need_handle_dfs_flag = true;
 	int len, i;
 	u32 cs_count;
+	u8 critical_update = 0;
 
 	if (!rdev->ops->channel_switch ||
 	    !(rdev->wiphy.flags & WIPHY_FLAG_HAS_CHANNEL_SWITCH))
@@ -11112,6 +11133,13 @@ static int nl80211_channel_switch(struct
 		}
 	}
 
+	if (csa_attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]) {
+		critical_update =
+			nla_get_u8(csa_attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]);
+		if (critical_update & NL80211_CU_ELEMID_ADDED)
+			params.beacon_csa_cu = 1;
+	}
+
 skip_beacons:
 	err = nl80211_parse_chandef(rdev, info, &params.chandef, wdev);
 	if (err)
@@ -11156,6 +11184,12 @@ skip_beacons:
 		if (err)
 			goto free;
 	}
+	if (info->attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]) {
+		critical_update =
+			nla_get_u8(info->attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]);
+		if (critical_update & NL80211_CU_ELEMID_MODIFIED)
+			params.beacon_after_cu = 1;
+	}
 
 	wdev_lock(wdev);
 	err = rdev_channel_switch(rdev, dev, &params, link_id);
@@ -16905,6 +16939,7 @@ static int nl80211_color_change(struct s
 	u16 offset;
 	int err;
 	unsigned int link_id = nl80211_link_id(info->attrs);
+	u8 critical_update = 0;
 
 	if (!rdev->ops->color_change)
 		return -EOPNOTSUPP;
@@ -16995,6 +17030,12 @@ static int nl80211_color_change(struct s
 		if (err)
 			goto out;
 	}
+	if (info->attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]) {
+		critical_update =
+			nla_get_u8(info->attrs[NL80211_ATTR_SET_CRITICAL_UPDATE]);
+		if (critical_update & NL80211_CU_ELEMID_ADDED)
+			params.elemid_added = 1;
+	}
 
 	wdev_lock(wdev);
 	err = rdev_color_change(rdev, dev, &params, link_id);
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -986,6 +986,7 @@ struct ieee80211_link_data_ap {
 
 	/* to be used after channel switch. */
 	struct cfg80211_beacon_data *next_beacon;
+	bool after_beacon_cu;
 };
 
 struct ieee80211_link_data {
